oak_client/src/attestation.rs

```diff
@@ -0,0 +1,156 @@
+        .context("Couldn't create TLS config")?;
+    Channel::builder(uri.clone())
+        .tls_config(tls_config)
+        .context("Couldn't create TLS configuration")?
+        .connect()
+        .await
+        .context("Couldn't connect to Oak Application")
+}
+
+/// Create client TLS configuration with a custom X.509 certificate verifier.
+/// [`root_tls_certificate`] should contain only a single PEM encoded certificate.
+/// https://tools.ietf.org/html/rfc1421
+pub fn create_tls_attestation_config(
+    root_tls_certificate: &[u8],
+    tee_quote: &str,
+) -> anyhow::Result<ClientTlsConfig> {
+    let mut config = rustls::ClientConfig::new();
+
+    // Enable HTTP/2.
+    const ALPN_H2: &str = "h2";
+    config.set_protocols(&[Vec::from(&ALPN_H2[..])]);
+
+    // Add root TLS certificate.
+    let mut cc_reader = std::io::BufReader::new(&root_tls_certificate[..]);
+    let certs = rustls::internal::pemfile::certs(&mut cc_reader)
+        .map_err(|error| anyhow!("Couldn't parse TLS certificate: {:?}", error))?;
+    if certs.len() != 1 {
+        return Err(anyhow!("Only a single PEM certificate should be provided"));
+    }
+    config
+        .root_store
+        .add(&certs[0])
+        .context("Couldn't add root certificate")?;
+
+    // Set custom X.509 certificate verifier.
+    config
+        .dangerous()
+        .set_certificate_verifier(std::sync::Arc::new(RemoteAttestationVerifier::new(
+            tee_quote,
+        )));
+
+    Ok(ClientTlsConfig::new().rustls_client_config(config))
+}
+
+/// Custom X.509 certificate verifier that checks TEE quotes in the X.509 TEE extension.
+struct RemoteAttestationVerifier {
+    tee_quote: String,
+}
+
+impl RemoteAttestationVerifier {
+    fn new(tee_quote: &str) -> Self {
+        Self {
+            tee_quote: tee_quote.to_string(),
+        }
+    }
+
+    fn check_tee_quote(&self, extension: &str) -> anyhow::Result<()> {
+        let pattern = r"^Quote:(.*)$";
+        let re = regex::Regex::new(&pattern).expect("Incorrect regular expression");
+        let tee_quotes = re.captures(&extension).context("Couldn't find TEE quote")?;
+
+        // `0` capture group corresponds to the whole pattern, while `1` corresponds to the first
+        // match.
+        // https://docs.rs/regex/1.4.3/regex/struct.Captures.html#method.get
+        if self.tee_quote == tee_quotes.get(1).map_or("", |m| m.as_str()) {
+            Ok(())
+        } else {
+            Err(anyhow!("Incorrect TEE quote"))
+        }
+    }
+}
+
+impl rustls::ServerCertVerifier for RemoteAttestationVerifier {
+    fn verify_server_cert(
+        &self,
+        roots: &rustls::RootCertStore,
+        certs: &[rustls::Certificate],
+        hostname: webpki::DNSNameRef,
+        ocsp: &[u8],
+    ) -> Result<rustls::ServerCertVerified, rustls::TLSError> {
+        if certs.len() != 1 {
+            return Err(rustls::TLSError::General(
+                "Only a single certificate should be provided".to_string(),
```

Comment:
```suggestion
                "Only a single certificate must be provided".to_string(),
```

Reply:
Maybe also return the actual number of certs found, it will help debugging. Could you also comment on why this is the case? Is it because we expect a self-signed cert? Or is it a single one even for CA-signed certs?